### PR TITLE
fix: brake temperature sometime starts at zero

### LINKED ISSUE #8153: Brake temperature in SD WHEEL page is sometimes 0°C at start and not ambiant temperature
### Aircraft Version

Development

### Build info

```json
{
    "built": "2023-08-03T20:10:35+00:00",
    "ref": "HEAD",
    "sha": "65308bc4fd1d2ddb859a404320934dd32e349fa9",
    "actor": "lukecologne <lukasmichels00@gmail.com>",
    "event_name": "manual",
    "pretty_release_name": "master:65308bc4",
    "version": "v0.11.0-dev.65308bc"
}
```

### Describe the bug

Sometimes when you start cold & dark or on runway, the brake temperatures shown on the SD Wheel page is 0°C. 

### Expected behavior

It should be ambiant temperature.

### Steps to reproduce

Start on the runway with ambiant temperature > 0°C. Open the WHEEL page on SD; it shows 0°C brake temperature.
I reproduce everytime but some people don't.

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

_No response_

### PR DESCRIPTION
**Pull request by pluce**
Fixes #8153

## Summary of Changes
Rewritten the brake temperature system and fix the original bug by initializing to ambiant temperature. Also, light up all "hottest brake" indication when they are more than one "hottest brake" (cf. [discussion with Tolip](https://discord.com/channels/738864299392630914/738864300818432023/1136921224665116682))

- [x] Need to add some documentation

## Testing instructions

- "Zero temperature" fix:

1.   try starting on runway or cold&dark on multiple airports and meteo conditions and check brakes temperature on SD WHEEL page: it should never be 0 if ambiant temperature exceeds 5 degrees.

- "All hottest brake light up":

1. Start on runway, throttle up then brake hard until multiple brakes exceeds 100°. If they are many with the same temperature, they should all have the green arc over the number

- Non regression tests:

1. After heating up the brakes as above, let the aircraft without input and it should slowly cool down the brakes
2. After heating up the brakes as above, let the aircraft without input, push the Brake Fan button and it should cool down the brakes faster 

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 1]
```diff
@@ -12,16 +12,26 @@ import { Spoilers } from '../../Common/Spoilers';
 
 import '../../Common/CommonStyles.scss';
 
+const roundTemperature = (rawTemp: number) : number => {
+    return Math.max(0, Math.round(rawTemp / 5) * 5)
```
Comment (BlueberryKing) [inline fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx:16]:
```suggestion
    return Math.max(0, Math.round(rawTemp / 5) * 5)
```

File: fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 2]
```diff
@@ -12,16 +12,26 @@ import { Spoilers } from '../../Common/Spoilers';
...
+
+    const roundedTemperatures = [
+        roundTemperature(rawTempBrake1),
+        roundTemperature(rawTempBrake2),
+        roundTemperature(rawTempBrake3),
+        roundTemperature(rawTempBrake4),
+    ]
 
-    const maxTemperatureIndex = temperatures.reduce((maxIndex, element, index, array) => (element > array[maxIndex] ? index : maxIndex), 0);
+    const maxTemperature = roundedTemperatures.reduce((maxTemp, element) => (element > maxTemp ? element : maxTemp), 0);
```
Comment (BlueberryKing) [inline fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx:34]:
```suggestion
    const maxTemperature = roundedTemperatures.reduce((maxTemp, element) => Math.max(maxTemp, element), 0);
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 3]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
-        const currentBrakeRight = SimVar.GetSimVarValue("BRAKE RIGHT POSITION", "position 32k");
 
-        const ambientTemperature = Simplane.getAmbientTemperature();
-        const airspeed = SimVar.GetSimVarValue("AIRSPEED TRUE", "Meters per second");
+        // First, the heat up process
 
-        const wheelSet1Rpm = SimVar.GetSimVarValue("WHEEL RPM:1", "number");
-        const wheelSet2Rpm = SimVar.GetSimVarValue("WHEEL RPM:2", "number");
+        // Don't need to check for spinning wheel of applied brakes, it's handled by calculateHeatUpPerSecond()
+        const heatUpFactor = secondRatio * getRandomArbitrary(0.5, 1.5) * this.gearStatus.calculateHeatUpPerSecond() ;
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:267]:
```suggestion
        const heatUpFactor = secondRatio * getRandomArbitrary(0.5, 1.5) * this.gearStatus.calculateHeatUpPerSecond();
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 4]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
+        this.brakeTemp += heatUpFactor;
+        this.reportedBrakeTemp += heatUpFactor;
 
-        const wheelsAreSpinning = wheelSet1Rpm > WHEEL_RPM_THRESHOLD || wheelSet2Rpm > WHEEL_RPM_THRESHOLD;
+        // Then the cooldown process
+        const deltaAmbiant = this.brakeTemp - ambiantTemperature;
+        // Cooldown from convection
+        if (Math.abs(deltaAmbiant) > MIN_TEMP_DELTA) {
+            const deltaTempFactor = 1 + Math.pow(deltaAmbiant, 2) * BASE_HEAT_DIFFERENTIAL_FACTOR * this.brakeFan.fanDifferentialFactor();
+            const brakeCoolDown = secondRatio * getRandomArbitrary(0.8, 1.2) * this.calculateDeltaCoolDown(deltaAmbiant, airspeed,  deltaTempFactor);
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:276]:
```suggestion
            const brakeCoolDown = secondRatio * getRandomArbitrary(0.8, 1.2) * this.calculateDeltaCoolDown(deltaAmbiant, airspeed, deltaTempFactor);
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 5]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
+            this.brakeTemp -= brakeCoolDown;
+            this.reportedBrakeTemp -= brakeCoolDown;
+        }
 
-        const anyBrakePressed = currentBrakeLeft > 0 || currentBrakeRight > 0;
+        if(this.brakeFan.shouldBeOn()) {
+            // When fan is on, it will cool the probe faster than the brakes
+            const probeTargetTemp = ambiantTemperature + deltaAmbiant / 2.0
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:283]: Semicolon
```suggestion
            const probeTargetTemp = ambiantTemperature + deltaAmbiant / 2.0;
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 6]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
+        this.brakes = [
+            new A32NX_OneBrakeTemp(0, "L:A32NX_BRAKE_TEMPERATURE_1", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_1", this.gearLeft, this.brakeFan),
+            new A32NX_OneBrakeTemp(1, "L:A32NX_BRAKE_TEMPERATURE_2", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_2", this.gearLeft, this.brakeFan),
+            new A32NX_OneBrakeTemp(2, "L:A32NX_BRAKE_TEMPERATURE_3", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_3", this.gearLeft, this.brakeFan),
+            new A32NX_OneBrakeTemp(3, "L:A32NX_BRAKE_TEMPERATURE_4", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_4", this.gearLeft, this.brakeFan)
+        ]
+    }
 
-            const deltaHeatUpWheelSet2 = (_deltaTime / 1000) * this.calculateHeatUp(currentBrakeRight, wheelSet2Rpm);
+    init() { }
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:317]:
```suggestion

```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 7]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
-            }
+    /**
+     * Update loop
+     * Fail fast if simulation is not ready
+     * Get ambiant temperature and airspeed, then update gear/brake fan systems from simulation
+     * Then compute brake temperature evolution
+     * Finally compute if any brake can be considered HOT
+     * @param {number} _deltaTime
+     */
+    update(_deltaTime) {
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:326]: Note for this specific case as well as generally: We use underscores in front of argument names to indicate that the argument is not used in the function, i.e `_deltaTime`. In this case, it does appear to be used however, so I'd change the name to `deltaTime`.

I believe there are earlier occurences where the parameter is indeed unused. In those cases, I suggest you either remove the argument or change the name to `_deltaTime` :slight_smile: .

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 8]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
-            SimVar.SetSimVarValue(`L:A32NX_BRAKE_TEMPERATURE_${i + 1}`, "celsius", currentBrakeTemps[i]);
-            SimVar.SetSimVarValue(`L:A32NX_REPORTED_BRAKE_TEMPERATURE_${i + 1}`, "celsius", currentReportedBrakeTemps[i]);
-            if (currentReportedBrakeTemps[i] > 300) {
-                brakesHot = 1;
-            }
-        }
+        this.brakes.forEach((brake) => {
+            brake.update(_deltaTime, ambientTemperature, airspeed);
+            someHot = someHot || brake.hot()
+        })
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:342]: Just semicolons.
```suggestion
        this.brakes.forEach((brake) => {
            brake.update(_deltaTime, ambientTemperature, airspeed);
            someHot = someHot || brake.hot();
        });
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 9]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
+     * Update loop
+     * Fail fast if simulation is not ready
+     * Get ambiant temperature and airspeed, then update gear/brake fan systems from simulation
+     * Then compute brake temperature evolution
+     * Finally compute if any brake can be considered HOT
+     * @param {number} _deltaTime
+     */
+    update(_deltaTime) {
+        if(!isReady()) {
+            return
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:328]:
```suggestion
        if (!isReady()) {
            return;
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 10]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
-                currentBrakeTemps[0] += brakeHeatUp0;
-                currentBrakeTemps[1] += brakeHeatUp1;
-                currentReportedBrakeTemps[0] += brakeHeatUp0;
-                currentReportedBrakeTemps[1] += brakeHeatUp1;
-            }
+    constructor() {
+        this.initializedAmbientBrakeTemp = false;
+        this.gearLeft = new A32NX_GearStatus(0, "L:A32NX_GEAR_LEFT_POSITION", "BRAKE LEFT POSITION", "WHEEL RPM:1");
+        this.gearRight = new A32NX_GearStatus(1, "L:A32NX_GEAR_RIGHT_POSITION", "BRAKE RIGHT POSITION", "WHEEL RPM:2");
+        this.brakeFan = new A32NX_BrakeFan("L:A32NX_BRAKE_FAN", "L:A32NX_BRAKE_FAN_BTN_PRESSED", this.gearLeft)
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:307]:
```suggestion
        this.brakeFan = new A32NX_BrakeFan("L:A32NX_BRAKE_FAN", "L:A32NX_BRAKE_FAN_BTN_PRESSED", this.gearLeft);
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 11]
```diff
@@ -1,16 +1,165 @@
...
+        return this.brakePosition > 0;
+    }
+
+    /**
+     * Check if this gear's wheel is spinning
+     * It checks RPM against a threshold
+     * @returns {boolean} true if this gear's wheel is spinning
+     */
+    wheelSpinning() {
+        return this.wheelRPM > WHEEL_RPM_THRESHOLD
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:86]: Semicolon
```suggestion
        return this.wheelRPM > WHEEL_RPM_THRESHOLD;
```

**Review by BlueberryKing — CHANGES_REQUESTED** [review]
Very nice work! Sorry for all the nitpicking with formatting.

Small English thing I noticed, it's `ambient`, not `ambiant`. Maybe you can quickly change that everywhere.

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 12]
```diff
@@ -20,186 +169,178 @@ const SPEED_COOLDOWN_FACTOR = 0.00055;
...
+        this.brakes = [
+            new A32NX_OneBrakeTemp(0, "L:A32NX_BRAKE_TEMPERATURE_1", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_1", this.gearLeft, this.brakeFan),
+            new A32NX_OneBrakeTemp(1, "L:A32NX_BRAKE_TEMPERATURE_2", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_2", this.gearLeft, this.brakeFan),
+            new A32NX_OneBrakeTemp(2, "L:A32NX_BRAKE_TEMPERATURE_3", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_3", this.gearLeft, this.brakeFan),
+            new A32NX_OneBrakeTemp(3, "L:A32NX_BRAKE_TEMPERATURE_4", "L:A32NX_REPORTED_BRAKE_TEMPERATURE_4", this.gearLeft, this.brakeFan)
+        ]
+    }
 
-            const deltaHeatUpWheelSet2 = (_deltaTime / 1000) * this.calculateHeatUp(currentBrakeRight, wheelSet2Rpm);
+    init() { }
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:316]: This does not appear to be used anymore.
```suggestion
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js [hunk 13]
```diff
@@ -1,16 +1,159 @@
...
+        this.wheelRPMSimVar = wheelRPMSimVar;
+        this.wheelRPM = 0;
+    }
 
-const BASE_HEAT_UP_FACTOR = 0.003;
+    /**
+     * Update internal state from simulation values
+     * @param {number} deltaTime
+     */
+    update(deltaTime) {
```
Comment (BlueberryKing) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:46]: I think you can just remove this `deltaTime` parameter entirely.

Comment (tracernz) [reply fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:46]: One thing I hate about JS/TS is that you can have mismatching signatures like that. 😬

Comment (pluce) [reply fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_BrakeTemp.js:46]: Yes I think it's better with the param in order not to forget it's called every frame....

**Review by BlueberryKing — APPROVED** [review]
LGTM

**Comment by auroraisluna** [comment]
QA Report

Discord: alepouna
Object of testing: #8156
Tier of Testing : 1
Date : 19/08/2023

Testing Process:

- "Zero temperature" fix
   - Start on runway or cold&dark on multiple airports and meteo conditions and check brakes temperature on SD WHEEL page
   - [PASS] It is not 0 degrees if ambient temp is over 5 degrees. 

- "All hottest brake light up"
   - [PASS] Start on runway, throttle up then brake hard until multiple brakes exceeds 100°. If they are many with the same temperature, they should all have the green arc over the number

- Non regression tests
   - [PASS] After heating up the brakes as above, let the aircraft without input and it should slowly cool down the brakes
   - [PASS] After heating up the brakes as above, let the aircraft without input, push the Brake Fan button and it should cool down the brakes faster

Testing Results:
Passed

**Pull request merged**